pkg/events/events.go

```diff
@@ -58,3 +58,73 @@ func CreateNodeEventRecord(objectID, reason, message string) (*si.EventRecord, e
 func CreateQueueEventRecord(objectID, groupID, reason, message string) (*si.EventRecord, error) {
 	return createEventRecord(si.EventRecord_QUEUE, objectID, groupID, reason, message)
 }
+
+// emitting event to the shim for a reservation
+// the event will be sent to the node, the application and the allocation involved in the reservation
+func EmitReserveEvent(allocKey, appID, nodeID string) error {
+	return emitRequestAppAndNodeEvents(allocKey, appID, nodeID,
+		"Allocation %s from application %s reserved node %s", "AppReservedNode")
+}
+
+// emit an unreserve event for the allocation, the application and the node as well
+func EmitUnReserveEvent(allocKey, appID, nodeID string) error {
+	return emitRequestAppAndNodeEvents(allocKey, appID, nodeID,
+		"Allocation %s from application %s unreserved from node %s", "AppUnreservedNode")
+}
+
+// emit an unreserve event only to the node
+func EmitUnreserveEventForNode(allocKey, appID, nodeID string) error {
+	eventCache := GetEventCache()
+	if eventCache == nil {
+		// The event cache is disabled.
+		// It's not an error, let's return.
+		return nil
+	}
+	msg := fmt.Sprintf("Ask %s from application %s allocated to node %s", allocKey, appID, nodeID)
+
+	event, err := CreateNodeEventRecord(nodeID, "AppAllocatedReservedNode", msg)
+	if err != nil {
+		return err
+	}
+	eventCache.AddEvent(event)
+
+	return nil
+}
+
+// emit an allocation event to the node through the shim
+func EmitAllocatedReservedEvent(allocKey, appID, nodeID string) error {
+	return emitRequestAppAndNodeEvents(allocKey, appID, nodeID,
+		"Ask %s from application %s allocated to node %s", "AppAllocatedReservedNode")
+}
+
+func emitRequestAppAndNodeEvents(allocKey, appID, nodeID, msgFmt, reason string) error {
+	eventCache := GetEventCache()
+	if eventCache == nil {
+		// The event cache is disabled.
+		// It's not an error, let's return.
+		return nil
+	}
+	msg := fmt.Sprintf(msgFmt, allocKey, appID, nodeID)
+
+	event, err := CreateNodeEventRecord(nodeID, reason, msg)
+	if err != nil {
+		return err
```

Comment:
Generic question: if this event create fails does that also mean all next events should or must fail. Could we have a case that just the nodeID is an issue and should we send out the rest of the events in that case?
I think we need to be as smart as possible and sent as much as we can (i.e. fail only what really must fail)

Reply:
You're right. Modified that part, also adjusted the unit tests.